logbook-jaxrs/src/main/java/org/zalando/logbook/jaxrs/LocalRequest.java

```diff
@@ -1,28 +1,101 @@
 package org.zalando.logbook.jaxrs;
 
+import lombok.AllArgsConstructor;
 import org.zalando.logbook.HttpRequest;
 import org.zalando.logbook.Origin;
 
 import javax.annotation.Nullable;
 import javax.ws.rs.client.ClientRequestContext;
+import java.io.IOException;
 import java.nio.charset.Charset;
 import java.util.List;
 import java.util.Map;
 import java.util.Optional;
+import java.util.concurrent.atomic.AtomicReference;
 
 import static java.util.Optional.ofNullable;
 
+@AllArgsConstructor
 final class LocalRequest implements HttpRequest {
 
-    private final ClientRequestContext context;
+    private interface State {
+
+        default State with() {
+            return this;
+        }
+
+        default State without() {
+            return this;
+        }
+
+        default State buffer(final ClientRequestContext context) {
+            return this;
+        }
+
+        default byte[] getBody() {
+            return new byte[0];
+        }
+
+    }
+
+    private static final class Unbuffered implements State {
+
+        @Override
+        public State with() {
+            return new Offering();
+        }
+
+    }
+
+    private static final class Offering implements State {
+
+        @Override
+        public State without() {
+            return new Unbuffered();
+        }
+
+        @Override
+        public State buffer(final ClientRequestContext context) {
+            final TeeOutputStream stream = new TeeOutputStream(context.getEntityStream());
+            context.setEntityStream(stream);
+            return new Buffering(stream);
+        }
+
+    }
 
-    private TeeOutputStream stream;
-    private byte[] body;
+    @AllArgsConstructor
+    private static final class Buffering implements State {
+
+        private final TeeOutputStream stream;
+
+        @Override
+        public State without() {
+            return new Ignoring(stream);
+        }
+
+        @Override
+        public byte[] getBody() {
+            return stream.toByteArray();
+        }
+
+    }
+
+    @AllArgsConstructor
+    private static final class Ignoring implements State {
+
+        private final TeeOutputStream stream;
+
+        @Override
+        public State with() {
+            return new Buffering(stream);
+        }
 
-    public LocalRequest(final ClientRequestContext context) {
-        this.context = context;
     }
 
+    private final AtomicReference<State> state = new AtomicReference<>(new Unbuffered());
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/app/whiskeysierra/logbook/pullRequest?prid=4506178)